src/dune/file_tree.ml

```diff
@@ -139,25 +141,41 @@ end = struct
       |> Result.ok
 end
 
-module Dir = struct
+module Dir0 = struct
+  module Contents = struct
+    type t =
+      { files : String.Set.t
+      ; sub_dirs : (Sub_dirs.Status.t * Readdir.t) String.Map.t
```

Comment:
It seems odd to me to store the readdir of all the children here. Intuitively, we should only need to do the readdir when descending into the children directories.

Reply:
It's indeed odd but there's no way around it with our current semantics I think. The reason why we need to store the contents is that we want to make sure we never expose unreadable directories to callers. At this point, we might as well save the contents so that we don't have compute it again when actually creating the subdir. An alternative is to just memoize `Readdir.of_source` and not worry about how many times we call it.

Reply:
I see. That's unfortunate.Another option would be to expose unreadable directories as empty directories. What do you think?

Reply:
I remember considering this when I first fixed the crashing on unreadable dirs. I can't recall the reasons why I didn't do it the way you're suggesting now. Perhaps because it could break users subtly somehow?

I'm happy to make that change but let's do it another PR. Here I'd like to preserve the current behavior as much as possible.

Reply:
Yh not sure. Though, it kind of matches the usual behaviour of the FS: you can list unreadable directories but you can't descend into them. So this change seems good to me.

Doing it in another PR seems fine, but at the same time it leaves master in a slightly unconfortable and temporary state. Given that we are close to the 2.0.0 release, that's not great :/ What about doing a first PR to switch the behaviour in master, and then rebase this PR on top?

Reply:
I just made the change in this PR. Turns out it wasn't so bad.